File: Cpp/Binary_Search.cpp
#include<iostream> 
using namespace std;
int show(int a[], int n) {
	cout<<"Sorted array is: ";
	for(int i = 0; i < n; i ++){
		cout<<a[i]<<" ";
	}
}
int binary(int a[], int s, int e, int x) {
	int mid = (s + e)/2;
	
	if(x > a[mid]) {
		
		binary(a, mid, e, x);
	}
	else if(x < a[mid]) {
		
		binary(a, s, mid, x);
	}
	else if(x == a[mid]) {

		return mid;		
	}
}
int sort(int arr[], int n) {
	for(int i = 0; i < n; i++) {
		for(int j = i + 1; j < n; j ++) {
			if(arr[i]>= arr[j]) {
				int temp;
				temp = arr[i];
				arr[i] = arr[j];
				arr[j] = temp;
			}
		}
	}
	show(arr, 6);
}
int main() {
	int x;
	int arr[6] = {12,18,20,42,8,10};
	cout<<"Enter the element to search: ";
	cin>>x;	
	sort(arr, 6);
	int mid = binary(arr,0, 6, x);
	cout<<endl<<"Found it at "<<mid + 1<<" position";
}

File: Python/binary search.py
# BINARY SEARCH
# This search is performed in a "sorted" list. Searching of key begins from the middle index.
# Element is searched in the left sublist if element at the middle index is greater than the key.
# Element is searched in the right sublist if element at the middle index is lower than the key.
# This is a recursive process.
# Finally, function returns index of element if key is present. Returns '-1' if key is absent.


# Time Complexity
# Average Case: O(logn)
# Best Case: O(1) -> When key is present at the middle index.

def ans(array, x):
    low = 0
    high = 1
    while x > high:
        newLow = high+1
        high = high + (high - (low-1))
        low = newLow

    return binarySearch(array, x, 0, len(array)-1)


def binarySearch(array, x, low, high):

    while low <= high:

        mid = low + (high - low)//2

        if array[mid] == x:
            return mid

        elif array[mid] < x:
            low = mid + 1

        else:
            high = mid - 1

    return -1


array = [3, 4, 5, 6, 7, 8, 9]
x = 6

result = ans(array, x)

print(result)
